### PR TITLE
#167190533 Build User Profile

### PR DESCRIPTION
**Pull request by abdulfatai360**
#### What does this PR do?
- It implements the feature to enable user profile to be created upon registration
- It validates the inputs when a user is editing his/her profile

#### Description of Task proposed in this pull request?
- Update User model to reflect new attributes
- Create Hobby and Category models
- Create migrations for users, hobbies, and categories tables
- Create migrations for UserCategory and UserHobby join tables
- Write validations schema for inputs when a user edits his/her
- Update the User controller to enable profile creation
- Write tests for the new code changes

#### How should this be manually tested (Quality Assurance)?
- Register a new user by making a `POST` request to `api/v1/users/create`

#### What are the relevant pivotal tracker stories?
- [#167190533](https://www.pivotaltracker.com/story/show/167190533)

#### Any background context you want to add (Operations Impact)?
- There is a function in this PR that is dependent on a function from another PR. So that function needs to be hooked to the User controller before this PR can work as it should

#### What I have learned working on this feature
- I have had my first hands-on experience working with Sequelize ORM
- I have learned the concept of `join tables` in database management
- I have learned how to write unit tests for Sequelize model

#### Screenshots:
NA

### CODE REVIEW

**Review by henryade — DISMISSED** [review]

**Review by JuwonAbiola — DISMISSED** [review]

File: server/controllers/Followers.js [hunk 1]
```diff
@@ -57,19 +65,26 @@ class Followers {
...
       await UserFollower.destroy({
         where: {
-          userId: user.id,
+          userId: userToBeUnfollowed.id,
           followerId: id
         }
       });
-      serverResponse(res, 200, {
+
+      await Followers.UpdateUserFollowersAndFollowingsCount(
```
Comment (dinobi) [inline server/controllers/Followers.js]: `.UpdateFollowsCount`

File: server/controllers/Followers.js [hunk 2]
```diff
@@ -161,6 +176,25 @@ class Followers {
     }
     return false;
   }
+
+  /**
+   * @name UpdateUserFollowersAndFollowingsCount
```
Comment (dinobi) [inline server/controllers/Followers.js]: update this

File: server/controllers/Followers.js [hunk 3]
```diff
@@ -161,6 +176,25 @@ class Followers {
...
+   * @async
+   * @static
+   * @memberof Followers
+   * @param {Object} source the source user object
+   * @param {Object} target the target user object
+   * @returns {Null} Null object
+   */
+  static async UpdateUserFollowersAndFollowingsCount(source, target) {
```
Comment (dinobi) [inline server/controllers/Followers.js]: and this

File: server/controllers/Followers.js [hunk 4]
```diff
@@ -161,6 +176,25 @@ class Followers {
...
+    const followingsCount = await source.countAllFollowings();
+    const followersCount = await target.countAllFollowers();
+
+    await User.update({ followingsCount }, { where: { id: source.id } });
```
Comment (dinobi) [inline server/controllers/Followers.js:193]: Why do we have this attribute on a user?. Pls avoid the need to store counts/size of resources that can be generated by looking at a single table.

Comment (abdulfatai360) [reply server/controllers/Followers.js:193]: I didn't create a profile table, and I informed you about this when I wanted to start the task that I was only going to extend the existing user table.

As for storing the counts of the followers and followings, there is no provision for followers Count table, and even if there is, the only way I could get it is still by watching for when a user follows someone or is followed by someone, update the counts, and store it in a table.

@dinobi

File: server/controllers/Followers.js (hunk 3 above)
Comment (abdulfatai360) [reply server/controllers/Followers.js]: @dinobi I will change the method name accordingly.

File: server/controllers/Profiles.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+const { User } = models;
+
+/**
+ * @export
+ * @class Users
+ */
+class Profiles {
+  /**
+   * @name create
```
Comment (dinobi) [inline server/controllers/Profiles.js]: @abdulfatai360  Do we need this action?

**Review by chialuka — APPROVED** [review]

**Review by henryade — APPROVED** [review]

**Review by JuwonAbiola — APPROVED** [review]

**Pull request merged**